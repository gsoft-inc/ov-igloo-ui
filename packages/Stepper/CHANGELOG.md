# @igloo-ui/stepper

## 0.1.0

### Minor Changes

- 2cef14d: The initial release of the stepper component
