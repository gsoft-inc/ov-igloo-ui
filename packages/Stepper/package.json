{
  "name": "@igloo-ui/stepper",
  "version": "0.2.2",
  "main": "dist/Stepper.js",
  "module": "dist/Stepper.js",
  "types": "dist/Stepper.d.ts",
  "styles": "dist/stepper.css",
  "publishConfig": {
    "access": "public"
  },
  "repository": "git+https://github.com/gsoft-inc/ov-igloo-ui.git",
  "author": "Officevibe Inc.",
  "contributors": [
    "Vicky Comeau vicky.coemau@gsoft.com"
  ],
  "license": "Apache-2.0",
  "peerDependencies": {
    "react": ">= 16.8.6"
  },
  "files": [
    "dist"
  ],
  "scripts": {
    "build": "rollup -c rollup.config.js"
  },
  "dependencies": {
    "@igloo-ui/tokens": "^2.0.0",
    "@igloo-ui/tooltip": "^3.4.2",
    "@hopper-ui/tokens": "^3.1.1",
    "classnames": "^2.3.2",
    "react-aria": "^3.30.0"
  },
  "devDependencies": {
    "@igloo-ui/button": "^0.7.3"
  },
  "browserslist": [
    "> 1%",
    "last 2 versions",
    "IE >= 11"
  ]
}
