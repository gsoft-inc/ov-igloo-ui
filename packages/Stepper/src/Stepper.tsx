import * as React from "react";
import cx from "classnames";
import type { PressEvent } from "react-aria";

import Step from "./Step";

import "./stepper.scss";

export interface Step {
    /** The callback function that is called when the step is clicked */
    onClick?: ((index: number, e?: PressEvent) => void);
    /** The title for the step */
    title: string;
}

export interface StepperProps extends React.ComponentProps<"div"> {
    /** The class name for the stepper */
    className?: string;
    /**
   * If true, the steps after the current step will be clickable
   * and the user will be able to
   * navigate to any step in the sequence
   */
    clickableNextSteps?: boolean;
    /** The current step in the sequence */
    currentStep?: number;
    /** Add a data-test tag for automated tests */
    dataTest?: string;
    /** The steps in the sequence */
    steps: Step[];
}

const Stepper: React.FunctionComponent<StepperProps> = ({
    className,
    clickableNextSteps = false,
    currentStep = 0,
    dataTest,
    steps,
    ...rest
}: StepperProps) => {
    const classes = cx("ids-stepper", className);

    return (
        <div className={classes} data-test={dataTest} {...rest}>
            {steps.map((step, index) => {
                const isCurrent = index === currentStep;
                const isComplete = index < currentStep;
                const disabled = !clickableNextSteps && index > currentStep;

                return (
                    <React.Fragment key={`frag_${index.toString()}`}>
                        <Step
                            key={`step_${index.toString()}`}
                            title={step.title}
                            isComplete={isComplete}
                            isCurrent={isCurrent}
                            disabled={disabled}
                            onPress={(e: PressEvent) => {
                                step.onClick?.(index, e);
                            }}
                        />
                        {index < steps.length - 1 && (
                            <div
                                key={`step-divider_${index.toString()}`}
                                className="ids-stepper__divider"
                            />
                        )}
                    </React.Fragment>
                );
            })}
        </div>
    );
};

export default Stepper;
