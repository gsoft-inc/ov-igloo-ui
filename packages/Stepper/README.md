# Stepper

The Stepper component provides a visual representation of a multi-step process.

<Example is="custom" />

<ReferenceLinks is="custom" />

## Installation

To install `@igloo-ui/stepper` in your project, you will need to run the following command using [npm](https://www.npmjs.com/):

```bash
npm install @igloo-ui/stepper
```

If you prefer [Yarn](https://classic.yarnpkg.com/en/), use the following command instead:

```bash
yarn add @igloo-ui/stepper
```

## Usage

Then to use the component in your code just import it!

```jsx
import Button from '@igloo-ui/button';
import Stepper from '@igloo-ui/stepper';

const [currentStep, setCurrentStep] = React.useState(0);
const exampleSteps = [
  { title: 'Step 1', onClick: (index) => setCurrentStep(index) },
  { title: 'Step 2', onClick: (index) => setCurrentStep(index) },
  { title: 'Step 3', onClick: (index) => setCurrentStep(index) },
];

<Stepper
  currentStep={currentStep}
  style={{ flex: '1 1 auto' }}
  steps={exampleSteps}
/>
<Button
  onClick={() => setCurrentStep((prevCurrentStep) => prevCurrentStep + 1)}
  disabled={currentStep === exampleSteps.length - 1}
>
  Continue
</Button>;
```
