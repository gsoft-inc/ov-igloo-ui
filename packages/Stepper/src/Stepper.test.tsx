/**
  * @jest-environment jsdom
  */
import React from 'react';
import { fireEvent, render, screen } from '@testing-library/react';

import MockTooltip from '@igloo-ui/tooltip/src/__mocks__/Tooltip.mock';

import Stepper, {StepperProps} from './Stepper';

jest.mock('@igloo-ui/tooltip', () => ({
  __esModule: true,
  default: jest.fn(MockTooltip),
}));

const steps = [
  { title: 'Step 1' },
  { title: 'Step 2' },
  { title: 'Step 3' },
];

const setup = (props: StepperProps = {
  steps: steps
 }) => {
  return render(
    <Stepper dataTest="ids-stepper" {...props} />
  );
};

describe('Stepper', () => {
   test('It should render without errors', () => {
    setup();
     const wrapper = screen.getByTestId('ids-stepper');
     expect(wrapper).toBeInTheDocument();
   });

   test('It should render a snapshot', () => {
     const {asFragment} = setup();
     expect(asFragment()).toMatchSnapshot();
   });

   test('Renders the correct number of steps', () => {
    const {container} = setup();
    const stepElements = container.querySelectorAll('.ids-step');
    expect(stepElements.length).toBe(steps.length);
  });

  test('Marks the current step as active', () => {
    const currentStep = 1;
    setup({steps: steps, currentStep: currentStep});
    const activeStep = screen.getByLabelText(steps[currentStep].title);
    expect(activeStep).toHaveClass('ids-step--current');
  });

  test('Calls the step onClick callback when a step is clicked', () => {
    const onStepChange = jest.fn();
    const stepsWithClicks = [
      { title: 'Step 1', onClick: onStepChange },
      { title: 'Step 2', onClick: onStepChange },
      { title: 'Step 3', onClick: onStepChange },
    ];
    const {container} = setup({steps: stepsWithClicks, currentStep: 2});
    const stepElements = container.querySelectorAll('.ids-step');
    fireEvent.click(stepElements[0]);
    expect(onStepChange).toHaveBeenCalledWith(0);
  });

  test('Disables steps after the current step if clickableNextSteps is false', () => {
    const currentStep = 1;
    const {container} = setup({steps: steps, currentStep: currentStep, clickableNextSteps: false});
    const stepElements = container.querySelectorAll('.ids-step');
    for (let i = currentStep + 1; i < stepElements.length; i++) {
      expect(stepElements[i]).toBeDisabled();
    }
  });

  test('Enables steps after the current step if clickableNextSteps is true', () => {
    const currentStep = 1;

    const {container} = setup({steps: steps, currentStep: currentStep, clickableNextSteps: true});
    const stepElements = container.querySelectorAll('.ids-step');
    for (let i = currentStep + 1; i < stepElements.length; i++) {
      expect(stepElements[i]).not.toBeDisabled();
    }
  });
});
