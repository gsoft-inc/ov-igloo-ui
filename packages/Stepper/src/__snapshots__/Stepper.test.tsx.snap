// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Stepper It should render a snapshot 1`] = `
<DocumentFragment>
  <div
    class="ids-stepper"
    data-test="ids-stepper"
  >
    <span
      class="ids-tooltip__container ids-step__wrapper ids-step__wrapper--current"
      role="button"
      tabindex="0"
    >
      <button
        aria-label="Step 1"
        class="ids-step ids-step--current"
        tabindex="-1"
        type="button"
      />
    </span>
    <div
      class="ids-stepper__divider"
    />
    <span
      class="ids-tooltip__container ids-step__wrapper ids-step__wrapper--inactive"
      role="button"
      tabindex="0"
    >
      <button
        aria-label="Step 2"
        class="ids-step ids-step--inactive"
        disabled=""
        tabindex="0"
        type="button"
      />
    </span>
    <div
      class="ids-stepper__divider"
    />
    <span
      class="ids-tooltip__container ids-step__wrapper ids-step__wrapper--inactive"
      role="button"
      tabindex="0"
    >
      <button
        aria-label="Step 3"
        class="ids-step ids-step--inactive"
        disabled=""
        tabindex="0"
        type="button"
      />
    </span>
  </div>
</DocumentFragment>
`;
